src/graphql/graphql-api.php

```diff
@@ -0,0 +1,231 @@
+				$sourced_block['innerBlocks'] = self::flatten_inner_blocks( $sourced_block );
+			}
+
+			if ( isset( $sourced_block['attributes'] ) && ! isset( $sourced_block['attributes'][0]['name'] ) ) {
+				$sourced_block['attributes'] = array_map(
+					function ( $name, $value ) {
+						return [
+							'name'  => $name,
+							'value' => $value,
+						];
+					},
+					array_keys( $sourced_block['attributes'] ),
+					array_values( $sourced_block['attributes'] )
+				);
+			}
+		}
+
+		return $sourced_block;
+	}
+
+	/**
+	 * Flatten the inner blocks, no matter how many levels of nesting is there.
+	 *
+	 * @param array $inner_blocks the inner blocks in the block.
+	 * @param array $flattened_blocks the flattened blocks that's built up as we go through the inner blocks.
+	 * 
+	 * @return array
+	 */
+	public static function flatten_inner_blocks( $inner_blocks, $flattened_blocks = [] ) {
+		if ( ! isset( $inner_blocks['innerBlocks'] ) ) {
+			array_push( $flattened_blocks, $inner_blocks );
+		} else {
+			$inner_blocks_copy = $inner_blocks['innerBlocks'];
+			unset( $inner_blocks['innerBlocks'] );
+			if ( isset( $inner_blocks['parentId'] ) ) {
+				array_push( $flattened_blocks, $inner_blocks );
+			}
+			foreach ( $inner_blocks_copy as $inner_block ) {
+				$flattened_blocks = self::flatten_inner_blocks( $inner_block, $flattened_blocks );
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockDataAttribute',
+			[
+				'description' => 'Block data attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => 'String',
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => 'String',
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual inner block, with the above attribute.
+		register_graphql_type(
+			'InnerBlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'         => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'parentId'   => [
+						'type'        => 'String',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. This will match the ID of the block',
+					],
+					'name'       => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes' => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
+			[
+				'description' => 'Block data',
```

Comment:
@ingeniumed Is there a way we can combine `BlockData` and `InnerBlockData` so that they're one datatype? It seems more intuitive that there's a single type, where a root `BlockData` has an empty parent.

Reply:
One more follow-up: am I correct that we're returning a flattened array of blocks that can be reconstructed by following `id`/`parentId` references? If that's the case, can we leave `InnerBlockData`/the `innerBlocks` attribute out of the GraphQL definitions, and rely on post-processing to make a block tree instead? It seems like there's a hybrid going on right now that feels unusual.